go/proxyd/backend.go

```diff
@@ -199,51 +306,219 @@ func (b *Backend) doForward(rpcReq *RPCReq) ([]byte, error) {
 
 type BackendGroup struct {
 	Name     string
-	backends []*Backend
-	i        int64
+	Backends []*Backend
 }
 
 func (b *BackendGroup) Forward(rpcReq *RPCReq) (*RPCRes, error) {
-	var outRes *RPCRes
-	for _, back := range b.backends {
+	rpcRequestsTotal.Inc()
+
+	for _, back := range b.Backends {
 		res, err := back.Forward(rpcReq)
-		if err == ErrBackendOffline {
+		if errors.Is(err, ErrMethodNotWhitelisted) {
+			return nil, err
+		}
+		if errors.Is(err, ErrBackendOffline) {
 			log.Debug("skipping offline backend", "name", back.Name)
 			continue
 		}
+		if errors.Is(err, ErrBackendOverCapacity) {
+			log.Debug("skipping over-capacity backend", "name", back.Name)
+			continue
+		}
 		if err != nil {
 			log.Error("error forwarding request to backend", "err", err, "name", b.Name)
 			continue
 		}
-		outRes = res
-		break
+		return res, nil
+	}
+
+	return nil, ErrNoBackends
+}
+
+func (b *BackendGroup) ProxyWS(clientConn *websocket.Conn) (*WSProxier, error) {
+	for _, back := range b.Backends {
+		proxier, err := back.ProxyWS(clientConn)
+		if errors.Is(err, ErrBackendOffline) {
+			log.Debug("skipping offline backend", "name", back.Name)
+			continue
+		}
+		if errors.Is(err, ErrBackendOverCapacity) {
+			log.Debug("skipping over-capacity backend", "name", back.Name)
+			continue
+		}
+		if err != nil {
+			log.Warn("error dialing ws backend", "name", back.Name, "err", err)
+			continue
+		}
+		return proxier, nil
+	}
+
+	return nil, ErrNoBackends
+}
+
+func calcBackoff(i int) time.Duration {
+	jitter := float64(rand.Int63n(250))
+	ms := math.Min(math.Pow(2, float64(i))*1000+jitter, 10000)
+	return time.Duration(ms) * time.Millisecond
+}
+
+type WSProxier struct {
+	backend     *Backend
+	clientConn  *websocket.Conn
+	backendConn *websocket.Conn
+}
+
+func NewWSProxier(backend *Backend, clientConn, backendConn *websocket.Conn, ) *WSProxier {
+	return &WSProxier{
+		backend:     backend,
+		clientConn:  clientConn,
+		backendConn: backendConn,
+	}
+}
+
+func (w *WSProxier) Proxy() error {
+	errC := make(chan error, 2)
```

Comment:
Is not closing this channel a memory leak?

Reply:
Nope! We actually don't want to close this channel. It's got two buffer slots so that closing either the client of backend end of the connection will close the other. If we close it, it'll cause the application to panic when the side that didn't close shuts down.